### PR TITLE
[staking] Cannot withdraw endorsement if it is withdrawing

### PR DESCRIPTION
**Pull request by envestcc**
# Description
PR includes the following changes:
- refactor endorsement validation 
- cannot withdraw endorsement if it is withdrawing
- add tests for endorsement feature

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/staking/bucket_validation.go [hunk 1]
```diff
@@ -77,24 +77,27 @@ func validateBucketSelfStake(featureCtx protocol.FeatureCtx, csm CandidateStateM
...
+	var status EndorsementStatus
+	if isEndorsed {
+		status = Endorsed | UnEndorsing
+	} else {
+		status = WithoutEndorsement | EndorseExpired
+	}
+	return validateBucketEndorsementByStatus(esm, bucket, status, height)
+}
+
+func validateBucketEndorsementByStatus(esm *EndorsementStateManager, bucket *VoteBucket, status EndorsementStatus, height uint64) ReceiptError {
```
Comment (CoderZhi) [inline action/protocol/staking/bucket_validation.go:89]: Efficiency is not the first thing to consider in this case, thus this refactor is not suggested.

File: action/protocol/staking/bucket_validation.go [hunk 2]
```diff
@@ -77,24 +77,27 @@ func validateBucketSelfStake(featureCtx protocol.FeatureCtx, csm CandidateStateM
...
-			return &handleError{
-				err:           errors.New("endorse bucket is expired"),
-				failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketType,
-			}
-		}
-		if !isEndorsed && status != EndorseExpired {
+		if !status.Contain(endorse.Status(height)) {
 			return &handleError{
-				err:           errors.New("bucket is already endorsed"),
+				err:           errors.Errorf("bucket endorsement is not in status %d", status),
```
Comment (CoderZhi) [inline action/protocol/staking/bucket_validation.go:95]: it is very confusing especially in this return error.

File: action/protocol/staking/bucket_validation_test.go [hunk 3]
```diff
@@ -118,14 +118,14 @@ func TestValidateBucket(t *testing.T) {
 		// endorsed bucket
 		r.NoError(esm.Put(bktIdx, &Endorsement{ExpireHeight: endorsementNotExpireHeight}))
 		r.Nil(validateBucketEndorsement(esm, bkt, true, blkHeight))
-		r.ErrorContains(validateBucketEndorsement(esm, bkt, false, blkHeight), "bucket is already endorsed")
+		r.ErrorContains(validateBucketEndorsement(esm, bkt, false, blkHeight), "bucket endorsement is not in status 9")
```
Comment (CoderZhi) [inline action/protocol/staking/bucket_validation_test.go:121]: for example, `9` is not straightforward to understand

**Review by CoderZhi — CHANGES_REQUESTED** [review]

File: action/protocol/staking/bucket_validation.go (hunk 1 above)
Comment (envestcc) [reply action/protocol/staking/bucket_validation.go]: reverted

File: action/protocol/staking/bucket_validation.go (hunk 2 above)
Comment (envestcc) [reply action/protocol/staking/bucket_validation.go]: add `EndorsementStatus.String()` method to improve readability

**Review by CoderZhi — APPROVED** [review]

File: action/protocol/staking/bucket_validation.go [hunk 4]
```diff
@@ -76,35 +75,53 @@ func validateBucketSelfStake(featureCtx protocol.FeatureCtx, csm CandidateStateM
...
-			}
+func validateBucketWithEndorsement(esm *EndorsementStateManager, bucket *VoteBucket, height uint64) ReceiptError {
+	status, err := esm.Status(bucket.Index, height)
+	if err != nil {
+		return &handleError{
+			err:           err,
+			failureStatus: iotextypes.ReceiptStatus_ErrUnknown,
+		}
+	}
+	if status != Endorsed && status != UnEndorsing {
```
Comment (millken) [inline action/protocol/staking/bucket_validation.go:86]: Can be simplified `if status == EndorseExpired`

**Review by millken — APPROVED** [review]

**Pull request merged**